node/src/components/linear_chain_sync.rs

```diff
@@ -147,53 +321,91 @@ where
     ) -> Effects<Self::Event> {
         match event {
             Event::Start(init_peer) => {
-                match self.init_hash {
-                    None => {
+                match self.state {
+                    State::None | State::Done => {
                         // No syncing configured.
+                        trace!("Received `Start` event when in {:?} state.", self.state);
                         Effects::new()
                     }
-                    Some(init_hash) => {
-                        trace!(?init_hash, "Start synchronization");
+                    State::SyncingTrustedHash { trusted_hash, .. } => {
+                        trace!(?trusted_hash, "Start synchronization");
                         // Start synchronization.
-                        fetch_block(effect_builder, init_peer, init_hash)
+                        fetch_block_by_hash(effect_builder, init_peer, trusted_hash)
+                    }
+                    State::SyncingDescendants { .. } => {
+                        panic!("`Start` event received when in `SyncingDescendants` state.")
                     }
                 }
             }
-            Event::BlockExecutionDone(block_hash, block_height) => {
-                info!(
-                    ?block_hash,
-                    ?block_height,
-                    "Finished linear chain blocks execution."
-                );
-                Effects::new()
-            }
-            Event::GetBlockResult(block_hash, fetch_result) => match fetch_result {
+            Event::GetBlockHeightResult(block_height, fetch_result) => match fetch_result {
+                BlockByHeightResult::Absent => match self.random_peer(rng) {
+                    None => {
+                        // `block_height` not found on any of the peers.
+                        // We have synchronized all, currently existing, descendants of trusted
+                        // hash.
+                        effect_builder
+                            .immediately()
+                            .event(move |_| Event::StartDownloadingDeploys)
+                    }
+                    Some(peer) => fetch_block_at_height(effect_builder, peer, block_height),
+                },
+                BlockByHeightResult::FromStorage(block) => {
+                    self.state.block_downloaded(&block);
+                    // We should be checking the local storage for linear blocks before we
+                    // start syncing.
+                    trace!(%block_height, "Linear block found in the local storage.");
+                    // If we found the linear block in the storage it means we should have
+                    // all of its parents as well. If that's not
+                    // the case then we have a bug.
+                    effect_builder
+                        .immediately()
+                        .event(move |_| Event::StartDownloadingDeploys)
```

Comment:
That should happen b/c the `linear_chain` will be empty and it will transition to `SyncingDescendants*` state. More complete handling of the cases that reuse local state will be added in the appropriate ticket.

Reply:
Aren't we already in `SyncingDescendants` state here?
And it could be that before the restart we already had block 100, couldn't it?
Then we would get it from storage here and decide we're done, even though we're not, I think?

Reply:
Current state of this component if adjusted to the current state of the other components. Since there's no option for reusing local storage when restarting it's not reflected in the code.

Reply:
> Since there's no option for reusing local storage when restarting

Not sure I follow this.  Why can't we use local storage on restart?
Andreas' concern seems valid to me too.  If correct, it means the comment about about having all the parents is irrelevant too - it's the descendants we're interested in here.

Reply:
Not at the moment at least since there's no option to get the last block from the linear chain. And if I understand correctly, there's no option to read whole linear chain from storage either.

> If correct, it means the comment about about having all the parents is irrelevant too - it's the descendants we're interested in here.

_Here_ we are interested in parents (it's `GetBlockHashResult` section). We're interested in descendants later on.

Reply:
That's what I thought too.

Reply:
Done.